File: app/src/main/java/com/zzhoujay/markdown/style/MarkDownQuoteSpan.java
package com.zzhoujay.markdown.style;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.text.Layout;
import android.text.style.QuoteSpan;

/**
 * Created by zhou on 16-6-25.
 * 引用Span
 */
public class MarkDownQuoteSpan extends QuoteSpan {

    private static final int STRIPE_WIDTH = 15;
    private static final int GAP_WIDTH = 40;

    public MarkDownQuoteSpan() {
        super();
    }

    public MarkDownQuoteSpan(int color) {
        super(color);
    }

    @Override
    public void drawLeadingMargin(Canvas c, Paint p, int x, int dir, int top, int baseline, int bottom, CharSequence text, int start, int end,
                                  boolean first, Layout layout) {
        Paint.Style style = p.getStyle();
        int color = p.getColor();
        p.setStyle(Paint.Style.FILL);
        p.setColor(getColor());
        c.drawRect(x, top, x + dir * STRIPE_WIDTH, bottom, p);
        p.setStyle(style);
        p.setColor(color);
    }

    @Override public int getLeadingMargin(boolean first) {
        return STRIPE_WIDTH + GAP_WIDTH;
    }

}

File: markdown/src/main/java/com/fastaccess/markdown/spans/MarkDownQuoteSpan.kt
package com.fastaccess.markdown.spans

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.text.Layout
import android.text.style.LeadingMarginSpan
import com.fastaccess.github.extensions.toPx


/**
 * Created by zhou on 16-6-25.
 * 引用Span
 */
class MarkDownQuoteSpan(private val color: Int) : LeadingMarginSpan {

    private val paint = Paint()
    private val rect = Rect()


    override fun drawLeadingMargin(
        c: Canvas,
        p: Paint,
        x: Int,
        dir: Int,
        top: Int,
        baseline: Int,
        bottom: Int,
        text: CharSequence,
        start: Int,
        end: Int,
        first: Boolean,
        layout: Layout) {
        val width = 4.toPx()
        paint.set(p)
        paint.style = Paint.Style.FILL
        paint.color = color
        val left: Int
        val right: Int
        val l = x + dir * width
        val r = l + dir * width
        left = Math.min(l, r)
        right = Math.max(l, r)
        rect.set(left, top, right, bottom)
        c.drawRect(rect, paint)
    }

    override fun getLeadingMargin(first: Boolean) = 24.toPx()

}
